File: client-libraries/python/arkos_client/resources/system.py
"""
System resource for the Arkos client library.
"""

from .base import Resource


class SystemResource(Resource):
    """System resource."""

    def get_info(self):
        """
        Get system information.

        Returns:
            dict: System information
        """
        return self._request("GET", "/api/v1/system/info")

    def get_stats(self):
        """
        Get system statistics.

        Returns:
            dict: System statistics
        """
        return self._request("GET", "/api/v1/system/stats")

    def get_config(self):
        """
        Get system configuration.

        Returns:
            dict: System configuration
        """
        return self._request("GET", "/api/v1/system/config")

    def update_config(self, config_data, requires_restart=True):
        """
        Update system configuration.

        Args:
            config_data (dict): Configuration data
            requires_restart (bool, optional): Whether the update requires a restart

        Returns:
            dict: Response data
        """
        return self._request(
            "PUT",
            "/api/v1/config/set",
            json={"requires_restart": 1 if requires_restart else 0, **config_data},
        )

    def restart(self):
        """
        Restart the system.

        Returns:
            dict: Response data
        """
        return self._request("POST", "/api/v1/restart")

    def get_logs(self, service, **kwargs):
        """
        Get system logs.

        Args:
            service (str): Service name (arkos, nginx, go2rtc)
            **kwargs: Additional parameters (download, stream, start, end)

        Returns:
            str: Logs
        """
        return self._request("GET", f"/api/v1/logs/{service}", params=kwargs)

    def get_version(self):
        """
        Get version.

        Returns:
            str: Version
        """
        return self._request("GET", "/api/v1/version")

    def get_config_schema(self):
        """
        Get configuration schema.

        Returns:
            dict: Configuration schema
        """
        return self._request("GET", "/api/v1/config/schema.json")

    def get_raw_config(self):
        """
        Get raw configuration.

        Returns:
            dict: Raw configuration
        """
        return self._request("GET", "/api/v1/config/raw")

    def save_config(self, config_content, save_option):
        """
        Save configuration.

        Args:
            config_content (str): Configuration content
            save_option (str): Save option

        Returns:
            dict: Response data
        """
        return self._request(
            "POST",
            "/api/v1/config/save",
            params={"save_option": save_option},
            data=config_content,
            headers={"Content-Type": "text/plain"},
        )

    def get_labels(self, camera=None):
        """
        Get labels.

        Args:
            camera (str, optional): Camera name

        Returns:
            list: Labels
        """
        params = {}
        if camera:
            params["camera"] = camera
        return self._request("GET", "/api/v1/labels", params=params)

    def get_sub_labels(self, split_joined=None):
        """
        Get sub-labels.

        Args:
            split_joined (int, optional): Split joined sub-labels

        Returns:
            list: Sub-labels
        """
        params = {}
        if split_joined is not None:
            params["split_joined"] = split_joined
        return self._request("GET", "/api/v1/sub_labels", params=params)

    def get_recognized_license_plates(self, split_joined=None):
        """
        Get recognized license plates.

        Args:
            split_joined (int, optional): Split joined license plates

        Returns:
            list: Recognized license plates
        """
        params = {}
        if split_joined is not None:
            params["split_joined"] = split_joined
        return self._request("GET", "/api/v1/recognized_license_plates", params=params)

    def get_timeline(self, camera="all", limit=100, source_id=None):
        """
        Get timeline.

        Args:
            camera (str, optional): Camera name or "all"
            limit (int, optional): Maximum number of items to return
            source_id (str, optional): Source ID

        Returns:
            list: Timeline
        """
        params = {"camera": camera, "limit": limit}
        if source_id:
            params["source_id"] = source_id
        return self._request("GET", "/api/v1/timeline", params=params)

    def get_hourly_timeline(self, **kwargs):
        """
        Get hourly timeline.

        Args:
            **kwargs: Query parameters
                cameras (str, optional): Comma-separated list of cameras or "all"
                labels (str, optional): Comma-separated list of labels or "all"
                after (float, optional): Filter events after this timestamp
                before (float, optional): Filter events before this timestamp
                limit (int, optional): Maximum number of items to return
                timezone (str, optional): Timezone

        Returns:
            dict: Hourly timeline
        """
        return self._request("GET", "/api/v1/timeline/hourly", params=kwargs)

File: client-libraries/javascript/src/resources/system.js
import Resource from './base';

/**
 * System resource.
 */
class SystemResource extends Resource {
  /**
   * Get system information.
   * @returns {Promise<Object>} System information
   */
  async getInfo() {
    return this.get('/api/v1/system/info');
  }

  /**
   * Get system statistics.
   * @returns {Promise<Object>} System statistics
   */
  async getStats() {
    return this.get('/api/v1/system/stats');
  }

  /**
   * Get system configuration.
   * @returns {Promise<Object>} System configuration
   */
  async getConfig() {
    return this.get('/api/v1/system/config');
  }

  /**
   * Update system configuration.
   * @param {Object} configData - Configuration data
   * @param {boolean} requiresRestart - Whether the update requires a restart
   * @returns {Promise<Object>} Response data
   */
  async updateConfig(configData, requiresRestart = true) {
    return this.put('/api/v1/config/set', {
      data: { requires_restart: requiresRestart ? 1 : 0, ...configData },
    });
  }

  /**
   * Restart the system.
   * @returns {Promise<Object>} Response data
   */
  async restart() {
    return this.post('/api/v1/restart');
  }

  /**
   * Get system logs.
   * @param {string} service - Service name (arkos, nginx, go2rtc)
   * @param {Object} options - Additional parameters
   * @param {string} options.download - Download filename
   * @param {boolean} options.stream - Stream logs
   * @param {number} options.start - Start line
   * @param {number} options.end - End line
   * @returns {Promise<string>} Logs
   */
  async getLogs(service, options = {}) {
    return this.get(`/api/v1/logs/${service}`, { params: options });
  }

  /**
   * Get version.
   * @returns {Promise<string>} Version
   */
  async getVersion() {
    return this.get('/api/v1/version');
  }

  /**
   * Get configuration schema.
   * @returns {Promise<Object>} Configuration schema
   */
  async getConfigSchema() {
    return this.get('/api/v1/config/schema.json');
  }

  /**
   * Get raw configuration.
   * @returns {Promise<Object>} Raw configuration
   */
  async getRawConfig() {
    return this.get('/api/v1/config/raw');
  }

  /**
   * Save configuration.
   * @param {string} configContent - Configuration content
   * @param {string} saveOption - Save option
   * @returns {Promise<Object>} Response data
   */
  async saveConfig(configContent, saveOption) {
    return this.post('/api/v1/config/save', {
      params: { save_option: saveOption },
      data: configContent,
      headers: { 'Content-Type': 'text/plain' },
    });
  }

  /**
   * Get labels.
   * @param {string} camera - Camera name
   * @returns {Promise<Array>} Labels
   */
  async getLabels(camera) {
    const params = {};
    if (camera) {
      params.camera = camera;
    }
    return this.get('/api/v1/labels', { params });
  }

  /**
   * Get sub-labels.
   * @param {number} splitJoined - Split joined sub-labels
   * @returns {Promise<Array>} Sub-labels
   */
  async getSubLabels(splitJoined) {
    const params = {};
    if (splitJoined !== undefined) {
      params.split_joined = splitJoined;
    }
    return this.get('/api/v1/sub_labels', { params });
  }

  /**
   * Get recognized license plates.
   * @param {number} splitJoined - Split joined license plates
   * @returns {Promise<Array>} Recognized license plates
   */
  async getRecognizedLicensePlates(splitJoined) {
    const params = {};
    if (splitJoined !== undefined) {
      params.split_joined = splitJoined;
    }
    return this.get('/api/v1/recognized_license_plates', { params });
  }

  /**
   * Get timeline.
   * @param {string} camera - Camera name or "all"
   * @param {number} limit - Maximum number of items to return
   * @param {string} sourceId - Source ID
   * @returns {Promise<Array>} Timeline
   */
  async getTimeline(camera = 'all', limit = 100, sourceId) {
    const params = { camera, limit };
    if (sourceId) {
      params.source_id = sourceId;
    }
    return this.get('/api/v1/timeline', { params });
  }

  /**
   * Get hourly timeline.
   * @param {Object} options - Query parameters
   * @param {string} options.cameras - Comma-separated list of cameras or "all"
   * @param {string} options.labels - Comma-separated list of labels or "all"
   * @param {number} options.after - Filter events after this timestamp
   * @param {number} options.before - Filter events before this timestamp
   * @param {number} options.limit - Maximum number of items to return
   * @param {string} options.timezone - Timezone
   * @returns {Promise<Object>} Hourly timeline
   */
  async getHourlyTimeline(options = {}) {
    return this.get('/api/v1/timeline/hourly', { params: options });
  }
}

export default SystemResource;
